uefi-test-runner/build.py

```diff
@@ -162,25 +166,71 @@ def run_qemu():
-    for line in qemu.stdout:
-        # Strip ending and trailing whitespace + ANSI escape codes for analysis
-        stripped = ansi_escape.sub('', line.strip())
-
-        # Skip empty lines
-        if not stripped:
-            continue
-
-        # Print out the processed QEMU output to allow logging & inspection
-        print(stripped)
-
-    # Wait for QEMU to finish, then abort if that fails
-    status = qemu.wait()
-    if status != 0:
-        raise sp.CalledProcessError(cmd=cmd, returncode=status)
+    # Start QEMU
+    qemu = sp.Popen(cmd, stdin=sp.PIPE, stdout=sp.PIPE, universal_newlines=True)
+    try:
+        # Connect to the QEMU monitor
+        with open(monitor_input_path, mode='w') as monitor_input,                  \
+             open(monitor_output_path, mode='r') as monitor_output:
+            # Execute the QEMU monitor handshake, doing basic sanity checks
+            assert monitor_output.readline().startswith('{"QMP":')
+            print('{"execute": "qmp_capabilities"}', file=monitor_input, flush=True)
+            assert monitor_output.readline() == '{"return": {}}\n'
+
+            # Iterate over stdout...
+            for line in qemu.stdout:
+                # Strip ending and trailing whitespace + ANSI escape codes
+                # (This simplifies log analysis and keeps the terminal clean)
+                stripped = ansi_escape.sub('', line.strip())
+
+                # Skip lines which contain nothing else
+                if not stripped:
+                    continue
+
+                # Print out the processed QEMU output for logging & inspection
+                print(stripped)
+
+                # If the app requests a screenshot, take it
+                if stripped.startswith("SCREENSHOT: "):
+                    reference_name = stripped[12:]
+
+                    # Ask QEMU to take a screenshot
+                    monitor_command = '{"execute": "screendump", "arguments": {"filename": "screenshot.ppm"}}'
+                    print(monitor_command, file=monitor_input, flush=True)
+
+                    # Wait for QEMU's acknowledgement, ignoring events
+                    reply = json.loads(monitor_output.readline())
+                    while "event" in reply:
+                        reply = json.loads(monitor_output.readline())
+                    assert reply == {"return": {}}
+
+                    # Tell the VM that the screenshot was taken
+                    print('OK', file=qemu.stdin, flush=True)
+
+                    # Compare screenshot to the reference file specified by the user
+                    # TODO: Add an operating mode where the reference is created if it doesn't exist
+                    reference_file = WORKSPACE_DIR / 'uefi-test-runner' / 'screenshots' / (reference_name + '.ppm')
+                    assert filecmp.cmp('screenshot.ppm', reference_file)
+
+                    # Delete the screenshot once done
+                    os.remove('screenshot.ppm')
```

Comment:
The way this is done, the screenshot.ppm will remain around if the comparison fails. This is a primitive implementation of the TODO above, not sure if we need something cleaner right now.

Reply:
How about using a [`tempfile`](https://docs.python.org/3/library/tempfile.html), and opening it in the `with` line?

Reply:
1. I'm not sure if tempfile is the right abstraction here, as it insists on opening the file whereas I need QEMU to write into it first :) A temporary directory would work, but seems a bit overkill for a single file.
2. As discussed above, we may actually want to keep the screenshot.ppm for manual inspection if the automatic comparison fails.